edb/lib/ext/auth.edgeql

```diff
@@ -55,17 +55,44 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
     };
 
     create type ext::auth::EmailFactor extending ext::auth::Factor {
-        create required property email: str {
-            create delegated constraint exclusive;
-        };
+        create required property email: str;
         create property verified_at: std::datetime;
     };
 
     create type ext::auth::EmailPasswordFactor
         extending ext::auth::EmailFactor {
+        alter property email {
+            create constraint exclusive;
+        };
         create required property password_hash: std::str;
     };
 
+    create type ext::auth::WebAuthnFactor extending ext::auth::EmailFactor {
+        create required property user_handle: std::bytes {
```

Comment:
I don't think this is an issue, just something that confused me initially: I was wondering why there was an email and a user handle, until I realised webauthn uses user handle to mean user id.

Reply:
But maybe we should call this `user_id`? To me 'handle' sounds like a user provided name, instead of a generated id that we create for them.

Reply:
Yeah, I decided to use the WebAuthn name for this so that it wasn't expected to actually match an "id" in our own system: it's a randomly generated byte sequence. It _should_ be unique for each `LocalIdentity` such that if you use multiple WebAuthn authenticators with the same email (this is allowed and encouraged!) you have the same `user_handle`.

Some additional reading:
  - https://w3c.github.io/webauthn/#sctn-user-handle-privacy
  - https://w3c.github.io/webauthn/#user-handle

I'm open to renaming this, but wanted to at least give some thoughts on why I chose this name.

Reply:
> such that if you use multiple WebAuthn authenticators with the same email (this is allowed and encouraged!) you have the same user_handle

Don't these constraints prevent that though? https://github.com/edgedb/edgedb/pull/6791/files#diff-77c4b6c26c6397edcdda77f3d29f3f96a657838c3fbbba98312e61c790c0affaR81-R82
I assumed that was the reason for having two constraints instead of `create constraint exclusive on ((.email, .credential_id, .user_handle));`

Reply:
Ah, you're right: we shouldn't drop that second constraint. We want to avoid having more than one object with the same email and credential ID, but an email can have multiple user handles and multiple credential IDs. Will fix.

Reply:
Is `WebAuthn` actually an email factor?  Is email always guaranteed to be provided?

Reply:
> Is WebAuthn actually an email factor? Is email always guaranteed to be provided?

It requires some kind of identifier, which doesn't strictly have to be an email address, but we kind of married ourselves to email as our "handle" right? It would be weird if webauthn allowed you to use usernames but you couldn't use usernames and passwords.

Or maybe you're asking if the authenticator gives us an email? It does not: the user provides that through the form.

Reply:
Strictly speaking you could have no identifier at all, but that's not typical UX.